whats-that-typeclass-functor/article.md

```diff
@@ -0,0 +1,467 @@
+
+Functor provides methods for structure-preserving transformations.
+
+You’re already familiar with `fmap`. However, just any `fmap` is not enough for a correct Functor instance. Some Haskell typeclasses have laws — conditions that instances have to meet. Only this guarantees that the instance is correct.
+
+Functor, in particular, needs `fmap` to adhere to the following laws.
+
+---
+
+### Laws of Functor
+
+1. **Identity**
+
+    `fmap id == id` — applying `id` function to the wrapped value changes nothing.
+
+    Example:
+
+    `fmap id (Just 1) == id (Just 1) == Just 1`
+
+2. **Composition**
+
+    `fmap f . fmap g == fmap (f . g)` — applying `fmap`s sequentially is the same as applying `fmap` with the composition of functions.
+
+    Example:
+
+    `fmap (+1) (fmap (*2) (Just 1)) == fmap ((+1) . (*2)) (Just 1) == Just 3`
+
+
+---
+
+`fmap` is all you need to define a `Functor` instance. However, the `(<$)` operator is worth looking at too.
+
+### The `(<$)` operator
+
+Take a look at the type signature of `(<$)`:
+
+```haskell
+(<$) :: a -> f b -> f a
+```
+
+The operator takes a value of type `a`, a value of type `b` packed in a functor `f` – `f b` – and produces a functor `f a`. Basically, the operator packs the first argument’s value in the context of the second argument, throwing away the second value.
+
+Now, let’s try to guess the definition of `(<$)` using the provided description and the following examples.
+
+```haskell
+-- replace an `Int` value
+-- inside `Maybe Int`
+-- with another `Int`
+> 1 <$ Just 2
+Just 1
+> 1 <$ Nothing
+Nothing
+
+-- replace an `Int` value
+-- inside `Maybe Int`
+-- with a `String`
+> "abc" <$ Just 123
+Just "abc"
+> "abc" <$ Nothing
+Nothing
+
+-- replace each element
+-- of the list `[Int]`
+-- with an `Int`
+> 1 <$ []
+[]
+> 1 <$ [0]
+[1]
+> 1 <$ [1..5]
+[1,1,1,1,1]
+
+-- replace each element
+-- of the list `[Int]`
+-- with a `Char`
+-- Note: the resulted type
+-- is `String` as `type String = [Char]`
+> '1' <$ []
+""
+> '1' <$ [0]
+"1"
+> '1' <$ [1..5]
+"11111"
+```
+
+Pay attention to list examples. `y <$ [x1, x2, x3]` matches `[y, y, y]` but not `[y]`, as list’s inside is many values, not a single one. So that each element of the list is transformed, but not the single value is wrapped in `[]`.
+
+
+<details>
+<summary> The default definition of <code>(<$)</code>.</summary>
+
+```haskell
+x <$ f = fmap (const x) f
+```
+
+<hr>
+</details>
+
+Exactly! `(<$)` just runs `fmap` with `const` function.
+
+
+We got to know the basic components of Functor, and now it’s time to create our own instance of this type class!
+
+## Creating an instance of Functor
+
+Let’s see how you can implement your own instance of Functor. We’ll use the [`NonEmpty`](https://hackage.haskell.org/package/base-4.16.1.0/docs/Data-List-NonEmpty.html#t:NonEmpty) data type as an example.
```

Comment:
> We’ll use the [`NonEmpty`]

It would be nice to use a type with a normal constructor so that the example is more accessable.

Reply:
Could NonEmpty be just written as a sum type?

Reply:
Yea, I guess so. Something like

```
data NonEmpty a = NonEmpty { neHead :: a, neTail :: [a] }
```
Right?